src/game.py

```diff
@@ -3,34 +3,47 @@
 from src.player import Player
 
 class Game():
+    EXIT = "exit"
 
     def __init__(self, cli=Cli(), board=Board(), players=[Player(token="X"),
                                                           Player(token="O")]):
         self.cli = cli
         self.board = board
         self.players = players
+        self.exit = False
 
     def start(self):
         self.cli.welcome()
         self.play()
 
     def play(self):
         self.cli.display_board(self.board)
-        while not self.board.full():
+        while not self.game_is_over():
             self.turn()
+        self.handle_exit()
 
-        self.cli.goodbye()
-    
-    def turn(self, error=False):
-        if error: self.cli.invalid_move()
-
-        player_choice = self.cli.prompt_player_turn(self.current_player())
-        move = self.input_to_index(player_choice)
-        if self.board.valid_move(move):
+    def handle_exit(self):
+        if self.exit:
+            self.cli.handle_exit()
+        else:
+            self.cli.handle_game_end()
+
+    def game_is_over(self):
+        return self.exit or self.board.full()
+
+    def turn(self):
+        while True:
```

Comment:
Can you think of another condition to use to exit the while loop? Something more specific to the purpose of the loop? Can you think of a way to avoid using a while loop at all here, perhaps with a recursive solution?

Reply:
Thats pretty funny that you ask that actually. A recursive solution is what I started with, if you have some time to take a closer look at the previous commits and conversations I had with dave that should tell a pretty good story about how I arrived at this solution. 

On another note, I plan on making some major changes to how this method works but I need to address the design of the program first, refactoring this method further will likely be rolled into that restructuring of the design.

Reply:
Got it. I couldn't find those comments, but that sounds fine. But it does make me curious, since I don't know the reasoning behind using the while loop, why is the `turn` method itself recursive instead implemented in a while loop as well?

Reply:
It's a convo that I resolved above, the first on on this PR. Not sure if you're able to see them after theyve been resolved but I assume you can. 

The reasoning behind the while loop is that it allows us to approach the finding of valid input in such a way that we only have one method call to prompt_user on the stack at any given time. That method resolves and then within the while loop a new call initiates if the input wasnt valid. This approach allows us to avoid overflowing the stack without having to put some sort of arbitrary limit on the number of times a user can put invalid input without us forcing a "crash" or the program crashing due to a stack trace limitation. 

I'm not sure what you mean by `Why si the turn method itself recursive` I don't think it is recursive any longer, can you point me to the line that you're referring to?